src/arch/riscv/tlb.cc

```diff
@@ -267,9 +295,10 @@ TLB::createPagefault(Addr vaddr, BaseMMU::Mode mode)
 }
 
 Addr
-TLB::translateWithTLB(Addr vaddr, uint16_t asid, BaseMMU::Mode mode)
+TLB::translateWithTLB(Addr vaddr, uint16_t asid, Addr xmode,
+                      BaseMMU::Mode mode)
 {
-    TlbEntry *e = lookup(vaddr, asid, mode, false);
+    TlbEntry *e = lookup(getVPNFromVAddr(vaddr, xmode), asid, mode, false);
     assert(e != nullptr);
     return e->paddr << PageShift | (vaddr & mask(e->logBytes));
```

Comment:
This still is very odd to me.

Reply:
We *should* test them. I can go update the documentation, but from all of my tests `paddr` is the PPN. For huge pages, the PPN will be the aligned 4KiB page number, not the "super page" number.

So, we need to get the offset from the `vaddr` (based on the size of the page), and then *or* that with the upper bits of the address from the PPN. 

Does this make sense, or do you think there's still something wrong.

Overall, this code is *incredibly* confusing with many different (and contradictory) names for the same thing.

Reply:
Step 6 of the the translation in the spec says that if a page is a leaf page then the unused ppn parts must be 0. So I think you're correct.